### PR TITLE
Add IsHCons.cons

### PR DESCRIPTION
**Pull request by dwijnand**
See https://twitter.com/dwijnand/status/807742485342593024 for motivation.

### CODE REVIEW

**Comment by milessabin** [comment]
Could you add the corresponding method to `IsCCons` as well. I think it belongs in the same PR. You should be able to make the two completely symmetrical :-)

**Comment by milessabin** [comment]
They key insight is that the kind of product is preserved (`HList` and `Coproduct` are both "products" in this sense). Switching to `Tuple2` and `Either` for the signature might also be reasonable.

**Comment by dwijnand** [comment]
Well, that compiles. Is that what you meant? Does it make sense? I've not played around with Coproduct much..

Also excluded them from MiMa, is that right?

File: core/src/main/scala/shapeless/ops/coproduct.scala [hunk 1]
```diff
@@ -859,6 +859,7 @@ object coproduct {
 
     def head(c: C): Option[H]
     def tail(c: C): Option[T]
+    def cons(e: Either[H, T]): H :+: T
```
Comment (milessabin) [inline core/src/main/scala/shapeless/ops/coproduct.scala]: Close ... this should give us a `C` in the same way that `cons` on `IsHCons` gives us an `L`.

**Comment by dwijnand** [comment]
That's pretty harsh, @codecov-io..

**Comment by dwijnand** [comment]
@milessabin How does this look to you now? GTG?

**Comment by milessabin** [comment]
Yeah, that's about the size of it. Bung in some tests to keep codecov happy and I'll merge :-)

**Comment by dwijnand** [comment]
@milessabin green (including from codecov)

**Comment by milessabin** [comment]
LGTM! Many thanks ... merging :-)

**Pull request merged**